crawler/src/main/scala/io/scalac/newspaper/crawler/failure/FailureHandler.scala

```diff
@@ -1,17 +1,69 @@
 package io.scalac.newspaper.crawler.failure
 
-import akka.Done
-import akka.stream.scaladsl.Sink
-import com.typesafe.scalalogging.StrictLogging
-import io.scalac.newspaper.crawler.fetching.FetchingFlow.{URLFetchingException, URLFetchingResult, URLFetchingTimeout, URLNotFetched}
+import akka.actor.{ActorLogging, Props}
+import akka.persistence.{PersistentActor, SnapshotOffer}
+import io.scalac.newspaper.crawler.failure.FailureHandler._
+import io.scalac.newspaper.crawler.fetching.FetchingFlow.{URLFetched, URLFetchingResult}
+import io.scalac.newspaper.crawler.urls.URLsStore
 
-import scala.concurrent.Future
+import scala.collection.mutable
 
-trait FailureHandler extends StrictLogging {
 
-  def handleFailure: Sink[URLFetchingResult, Future[Done]] = Sink.foreach {
-    case URLNotFetched(url, responseCode, content) => logger.error(s"Unable to fetch: $url, response code: $responseCode, response body: $content")
-    case URLFetchingTimeout(url) => logger.error(s"Timeout when trying to fetch: $url")
-    case URLFetchingException(url, ex) => logger.error(s"Exception when trying to fetch url: $url", ex)
+object FailureHandler {
+  case class FailureReported(url: String, result: URLFetchingResult)
+
+  def props(urlsStore: URLsStore, config: FailureHandlerConfig = defaultConfig) = Props(new FailureHandler(urlsStore, config))
+
+  case class FailureHandlerConfig(failuresLimit: Int, snapshotInterval: Int)
+  val defaultConfig =  FailureHandlerConfig(3, 50)
+
+  type FailuresStore = mutable.Map[String, Seq[URLFetchingResult]]
+
+  case class GetFailures(url: String)
+  case class URLFailures(failures: Option[Seq[URLFetchingResult]])
+  case object Complete
+}
+
+class FailureHandler(urlsStore: URLsStore, config: FailureHandlerConfig) extends PersistentActor with ActorLogging {
+
+  private var failures: FailuresStore = mutable.Map.empty
+
+  override def persistenceId: String = "failure-handler"
+
+  override def receiveCommand: Receive = {
+    case GetFailures(url) => sender() ! URLFailures(failures.get(url))
+
+    case _:URLFetched => log.error("Received URLFetched in FailureHandler")
+
+    case f:URLFetchingResult =>
+      persist(FailureReported(f.url, f)) { failure =>
```

Comment:
I may be not understanding something here, but aren't you including the URL twice in `FailureReported`?